File: lib/cost-rates.ts
/**
 * Provider pricing for the voice pipeline.
 *
 * Rates are versioned (`RATES_SOURCED_AT`) and intentionally hard-coded
 * rather than pulled from an env-driven config — costs are part of the
 * product, not deployment plumbing, and changing them should require a
 * code review with a fresh date stamp. If a provider raises prices, the
 * dashboard cost numbers go silently wrong until someone updates this
 * file; the date stamp is the signal that they need refreshing.
 *
 * Sourcing notes (verified 2026-07-17 against each provider's own page):
 *  - Groq: https://groq.com/pricing
 *      openai/gpt-oss-120b: $0.15 / 1M input tokens, $0.60 / 1M output.
 *      WAS $0.75 output — that figure was stale; Groq's published output
 *      rate is $0.60. (llama-3.3-70b-versatile is decommissioned 2026-08-16;
 *      gpt-oss-120b is both cheaper and the only Groq tier with strict
 *      json_schema.)
 *  - ElevenLabs Flash v2.5: $0.05 / 1k characters synthesized.
 *      https://elevenlabs.io/pricing/api — "API usage is billed in US
 *      dollars, not credits ... Text to Speech $0.10 per 1,000 characters
 *      (Multilingual v2/v3) or $0.05 (Flash/Turbo)."
 *      We express the PUBLISHED PER-CHARACTER API RATE, not a subscription
 *      blended rate, because this pipeline calls the ElevenLabs API and API
 *      usage is billed directly in USD — subscription credits do not apply.
 *      WAS $0.18, derived from a broken blended-subscription calc. That calc
 *      was wrong twice over: (1) it used "100k chars" for the Creator plan,
 *      but Creator is $22 for 121k *credits*, and (2) Flash/Turbo consumes
 *      only 0.5 credits per character (Multilingual v2 is 1.0). Even done
 *      correctly the blended rate is $22 / (121k / 0.5) = $22 / 242k chars
 *      ≈ $0.09/1k — and it is moot here anyway, since API calls bill in USD
 *      at $0.05/1k, not against the subscription bucket. Turbo is deprecated;
 *      ElevenLabs recommends Flash over Turbo in all cases.
 *  - Deepgram Nova-3: https://deepgram.com/pricing
 *      $0.0048 / minute streaming, MONOLINGUAL. The pipeline pins
 *      language="en-US" (pipeline.py), so it runs Nova-3 monolingual, priced
 *      at $0.0048/min streaming (Pay As You Go). WAS $0.0058 — that is the
 *      Nova-3 *multilingual* streaming rate, the wrong variant for an en-US
 *      pipeline. Same class of bug as the nova-2/nova-3 mixup (see models.py).
 *  - LiveKit Cloud: https://livekit.com/pricing
 *      $0.0005 / participant ("WebRTC") minute. Build tier includes 5k
 *      minutes/mo free; the marginal WebRTC-minute rate beyond the free
 *      tier is $0.0005 (Ship) / $0.0004 (Scale). WAS $0.005 — a 10x
 *      decimal error. Two participants per session (agent + user), so
 *      cost-per-session-minute = 2 * 0.0005 = $0.001.
 *
 * All numbers expressed in USD. The Python mirror in
 * livekit-agent/src/interview_agent/cost_rates.py uses the same values
 * — if you change one, change both, and bump RATES_SOURCED_AT in both.
 */

export const RATES_SOURCED_AT = "2026-07-17" as const;

export const RATES = {
  groq: {
    "openai/gpt-oss-120b": {
      inputUsdPerMillion: 0.15,
      outputUsdPerMillion: 0.6,
    },
  },
  elevenlabs: {
    "eleven_flash_v2_5": {
      // Published per-character API rate. ElevenLabs bills API usage in USD
      // (not subscription credits) at $0.05/1k chars for Flash/Turbo. Flash
      // and Turbo are the same price; Flash is simply faster, and Turbo is
      // deprecated. See the header note for why this replaced the old
      // blended-subscription figure (which double-counted and didn't apply
      // to API billing anyway).
      usdPerThousandChars: 0.05,
    },
  },
  deepgram: {
    "nova-3": {
      // Monolingual streaming (Pay As You Go). The pipeline pins en-US.
      usdPerAudioMinute: 0.0048,
    },
  },
  livekit: {
    // Two participants (agent + user). Charged per WebRTC minute of *each*
    // participant's presence, so a 10-minute session costs
    // 2 * 10 * 0.0005 = $0.01. Build tier includes 5k min/mo free.
    usdPerParticipantMinute: 0.0005,
    participantsPerSession: 2,
  },
} as const;

// ---------------------------------------------------------------------------
// Conversion helpers
// ---------------------------------------------------------------------------

export interface CostBreakdown {
  /** USD total for the Groq LLM calls (input + output tokens). */
  groqUsd: number;
  /** USD total for ElevenLabs TTS character synthesis. */
  ttsUsd: number;
  /** USD total for Deepgram STT audio minutes. */
  sttUsd: number;
  /** USD total for LiveKit participant-minutes. */
  livekitUsd: number;
  /** Sum of the above. */
  totalUsd: number;
  /** Pricing snapshot date — surfaces stale rates to the dashboard. */
  ratesSourcedAt: typeof RATES_SOURCED_AT;
}

export function groqUsd(input: {
  inputTokens: number;
  outputTokens: number;
  model?: keyof typeof RATES.groq;
}): number {
  const r = RATES.groq[input.model ?? "openai/gpt-oss-120b"];
  if (!r) return 0;
  return (
    (input.inputTokens * r.inputUsdPerMillion) / 1_000_000 +
    (input.outputTokens * r.outputUsdPerMillion) / 1_000_000
  );
}

export function ttsUsd(input: {
  charactersCount: number;
  model?: keyof typeof RATES.elevenlabs;
}): number {
  const r = RATES.elevenlabs[input.model ?? "eleven_flash_v2_5"];
  if (!r) return 0;
  return (input.charactersCount * r.usdPerThousandChars) / 1_000;
}

export function sttUsd(input: {
  audioSeconds: number;
  model?: keyof typeof RATES.deepgram;
}): number {
  const r = RATES.deepgram[input.model ?? "nova-3"];
  if (!r) return 0;
  return (input.audioSeconds / 60) * r.usdPerAudioMinute;
}

export function livekitUsd(input: {
  sessionDurationSeconds: number;
}): number {
  const minutes = input.sessionDurationSeconds / 60;
  return (
    minutes *
    RATES.livekit.usdPerParticipantMinute *
    RATES.livekit.participantsPerSession
  );
}

/**
 * Roll a usage snapshot into a complete cost breakdown.
 *
 * Input shape mirrors the AgentSessionUsage struct emitted by
 * `session_usage_updated` on the Python side — flatten `model_usage`
 * into the four counts and pass them in. Unknown providers / models
 * resolve to $0.00 (rather than throw) so a session whose TTS
 * provider was swapped to Cartesia gracefully reports a partial cost
 * instead of failing.
 */
export function rollUpCost(input: {
  llmInputTokens: number;
  llmOutputTokens: number;
  ttsCharactersCount: number;
  sttAudioSeconds: number;
  sessionDurationSeconds: number;
}): CostBreakdown {
  const groq = groqUsd({
    inputTokens: input.llmInputTokens,
    outputTokens: input.llmOutputTokens,
  });
  const tts = ttsUsd({ charactersCount: input.ttsCharactersCount });
  const stt = sttUsd({ audioSeconds: input.sttAudioSeconds });
  const lk = livekitUsd({ sessionDurationSeconds: input.sessionDurationSeconds });
  return {
    groqUsd: groq,
    ttsUsd: tts,
    sttUsd: stt,
    livekitUsd: lk,
    totalUsd: groq + tts + stt + lk,
    ratesSourcedAt: RATES_SOURCED_AT,
  };
}

/** Format a USD value as "$0.14". Returns "$0.00" for non-finite input. */
export function formatUsd(usd: number): string {
  if (!Number.isFinite(usd) || usd < 0) return "$0.00";
  return `$${usd.toFixed(2)}`;
}

File: livekit-agent/src/interview_agent/cost_rates.py
"""Provider pricing — Python mirror of lib/cost-rates.ts.

Keep the two in sync. ``RATES_SOURCED_AT`` is the date stamp that
forces a refresh review when prices drift; bump it in *both* files
whenever a number here changes.

Model ids are imported from interview_agent.models, NOT retyped here. That is
deliberate: this table previously billed Deepgram nova-3 while the pipeline ran
nova-2, so every cost figure on the dashboard was wrong for a model that was
never running.

Sourcing notes (verified 2026-07-17; links + full rationale in the TS mirror):
  - Groq openai/gpt-oss-120b:      $0.15 / 1M in, $0.60 / 1M out
                                   (was $0.75 out — stale)
  - ElevenLabs eleven_flash_v2_5:  $0.05 / 1k characters (published API rate,
                                   billed in USD not credits; Flash/Turbo
                                   consume 0.5 credits/char, so the old $0.18
                                   blended-subscription figure was wrong)
  - Deepgram nova-3:               $0.0048 / audio minute (monolingual
                                   streaming; pipeline pins en-US. Was $0.0058,
                                   the multilingual rate — wrong variant)
  - LiveKit:                       $0.0005 / participant (WebRTC) minute
                                   (was $0.005, a 10x decimal error),
                                   2 participants/session
"""

from __future__ import annotations

from dataclasses import dataclass

RATES_SOURCED_AT = "2026-07-17"

# Groq DEFAULT_LLM_MODEL.
_GROQ_INPUT_USD_PER_MILLION = 0.15
_GROQ_OUTPUT_USD_PER_MILLION = 0.6

# ElevenLabs TTS_MODEL. Published per-character API rate ($0.05/1k), not a
# subscription blended rate — API usage bills in USD, not credits.
_TTS_USD_PER_THOUSAND_CHARS = 0.05

# Deepgram STT_MODEL. Monolingual streaming (pipeline pins en-US).
_STT_USD_PER_AUDIO_MINUTE = 0.0048

# LiveKit WebRTC minutes. Build tier includes 5k min/mo free.
_LIVEKIT_USD_PER_PARTICIPANT_MINUTE = 0.0005
_LIVEKIT_PARTICIPANTS_PER_SESSION = 2


@dataclass(frozen=True)
class CostBreakdown:
    """Dollar breakdown of one session. Field names match the TS shape
    so the Firestore round-trip preserves them verbatim."""

    groq_usd: float
    tts_usd: float
    stt_usd: float
    livekit_usd: float
    total_usd: float
    rates_sourced_at: str

    def to_firestore_dict(self) -> dict[str, float | str]:
        """Camel-case the keys to match the TS Session.estimatedCost shape."""
        return {
            "groqUsd": round(self.groq_usd, 6),
            "ttsUsd": round(self.tts_usd, 6),
            "sttUsd": round(self.stt_usd, 6),
            "livekitUsd": round(self.livekit_usd, 6),
            "totalUsd": round(self.total_usd, 6),
            "ratesSourcedAt": self.rates_sourced_at,
        }


def groq_usd(input_tokens: int, output_tokens: int) -> float:
    return (
        input_tokens * _GROQ_INPUT_USD_PER_MILLION / 1_000_000
        + output_tokens * _GROQ_OUTPUT_USD_PER_MILLION / 1_000_000
    )


def tts_usd(characters_count: int) -> float:
    return characters_count * _TTS_USD_PER_THOUSAND_CHARS / 1_000


def stt_usd(audio_seconds: float) -> float:
    return (audio_seconds / 60.0) * _STT_USD_PER_AUDIO_MINUTE


def livekit_usd(session_duration_seconds: float) -> float:
    minutes = session_duration_seconds / 60.0
    return (
        minutes
        * _LIVEKIT_USD_PER_PARTICIPANT_MINUTE
        * _LIVEKIT_PARTICIPANTS_PER_SESSION
    )


def roll_up_cost(
    *,
    llm_input_tokens: int,
    llm_output_tokens: int,
    tts_characters_count: int,
    stt_audio_seconds: float,
    session_duration_seconds: float,
) -> CostBreakdown:
    """Sum all four legs into one CostBreakdown."""
    g = groq_usd(llm_input_tokens, llm_output_tokens)
    t = tts_usd(tts_characters_count)
    s = stt_usd(stt_audio_seconds)
    lk = livekit_usd(session_duration_seconds)
    return CostBreakdown(
        groq_usd=g,
        tts_usd=t,
        stt_usd=s,
        livekit_usd=lk,
        total_usd=g + t + s + lk,
        rates_sourced_at=RATES_SOURCED_AT,
    )
